modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1763,13 +1765,154 @@ def sort_by(
+        if not isinstance(columns, list):
+            columns = [columns]
+        axis = Axis(axis)
+        # This if selects cases where we are either sorting by a single column that is not
+        # the index, or we are sorting by multiple columns, since in this case, we will have to
+        # shuffle the data before sorting.
+        if axis == Axis.ROW_WISE and not (
+            len(columns) == 1 and columns[0] == "__index__"
+        ):
+            if self.dtypes[columns[0]] == object:
+                # This means we are not sorting numbers, so we need our quantiles to not try
+                # arithmetic on the values.
+                method = "inverted_cdf"
+            else:
+                method = "linear"
+
+            def sample_func(df, A=100, k=0.05, q=0.1):
+                """
+                Sample the given partition.
+
+                This function is applied to each row-axis partition in parallel to select samples
+                from each. It samples using the following algorithm:
+                    * If there are <= 100 rows in the dataframe, we pick quantiles over the entire dataframe.
+                    * If there are 100 < # of rows <= 1900, we pick quantiles over the first 100 rows, plus a sample
+                    of 5% of the remaining rows.
+                    * If there are > 1900 rows, we pick quantiles over a sample of 10% of all of the rows.
+
+                These numbers are a heuristic. They were picked such that the size of the sample
+                in scenario 2 scales well. In other words, we picked A, k, and q such that:
+                    A + (len(df) - A)*k = q*len(df)
+                Where q is the proportion we sample in scenario 3, k is the proportion we sample
+                of remaining rows in scenario 2, and A is the threshold below which we just return
+                the entire dataframe, instead of sampling (scenario 1).
+                q = 0.1 and k = 0.05 were picked such that this formula holds for A = 100.
+                """
+                quantiles = [
+                    i / (NPartitions.get() * 2) for i in range(NPartitions.get() * 2)
+                ]
+                # Heuristic for a "small" df we will compute quantiles over entirety of.
+                if len(df) <= A:
+                    return np.quantile(df[columns[0]], quantiles, method=method)
+                # Heuristic for a "medium" df where we will include first 100 (A) rows, and sample
+                # of remaining rows when computing quantiles.
+                if len(df) <= A * (1 - k) / (1 - q):
+                    return np.quantile(
+                        np.concatenate(
+                            (
+                                df[columns[0]][:100].values,
+                                df[columns[0]][100:].sample(frac=k),
+                            )
+                        ),
+                        quantiles,
+                        method=method,
```

Comment:
This will break if the latest version of numpy isn't installed, but is necessary to be able to handle sorting objects that are not numeric.